### ISSUE TITLE
Allow custom, isolated classloaders for different JavaScript contexts

### ISSUE THREAD

**Issue by rbeckman-nextgen** [nextgenhealthcare/connect#3449]
Right now there's a single, global Rhino context factory, and it's used by all JavaScript contexts across the entire application. Since it's only initialized once, any custom libraries / etc. included in custom-lib currently require a server restart to be included in channels and connectors. Instead, the way we compile and execute JavaScript using Rhino should be re-architected so that custom classloader can be hot deployed without having to restart the server.

Also, when such a change happens, any place that uses JavaScript should have a way to detect those changes and automatically recompile its associated scripts using the new context factory.

Imported Issue. Original Details:
Jira Issue Key: MIRTH-3553
Reporter: narupley
Created: 2014-12-15T14:52:03.000-0800

**Comment by rbeckman-nextgen** [comment]
Revision 7494: The way Rhino context factories are created and used has been completely overhauled to allow isolated contexts with specific classloaders. This allows custom libraries to be included within Mirth Connect channels **without** a server restart. This applies not only to JavaScript contexts, but also to other places in Mirth Connect where custom libraries might be used, like the Web Service Listener, Database connectors, JMS connectors, and custom attachment handler. When a context factory change is detected, the associated script is automatically re-compiled with the new context (and new sealed/shared scope). A new ContextFactoryController has been added to take control of context factory creation/reloading; everything that needs a context factory will request it from the controller. The library resources added with MIRTH-3554 tie into the controller as well.

Imported Comment. Original Details:
Author: narupley
Created: 2014-12-15T15:12:20.000-0800

**Comment by rbeckman-nextgen** [comment]
I created several small classes, loading one of them into the custom-lib folder upon server restart. I added my additional library class and then refreshed my libraries. I verified I was able to use this new library without first having to restart my server

Imported Comment. Original Details:
Author: jaysenp
Created: 2015-01-19T17:55:21.000-0800

**Comment by rbeckman-nextgen** [comment]
This issue, and by association MIRTH-3554, will require more testing. Please coordinate with Nick to come up with a test plan.

Imported Comment. Original Details:
Author: wayneh
Created: 2015-01-19T18:18:36.000-0800

**Comment by rbeckman-nextgen** [comment]
To test this issue, I performed the following tasks:

I created 5 simple test classes and packaged them as individual jars. Two of the classes were added to the '[Default Resource]' entry (custom-lib by default) and the other three added to their own separate resource directories.
I created a channel with multiple destinations. Each destination was configured to use a separate resource except for the source connector which was configured to use all the resources. The last destination was configured to use no resources.
I verified, through the process of reconfiguring each destination during testing, that those with resources set were only able to use these resources and no others.
I verified that all resources flagged as global were accessible by global scripts, and conversely I verified that non-global resources weren't accessible.
I added some existing Mirth Connect libraries, such as slf4j, to the custom-lib folder to ensure nothing went seriously awry.
Several versions of an existing jar were added to test any conflicts. In each case, it seemed the resource that appeared first in the resources list was used instead of the most recent or oldest existing version present.
I finally deleted all my channels and resources and restored them from a backup. I verified that these channels worked and were still configured properly.

Imported Comment. Original Details:
Author: jaysenp
Created: 2015-01-20T12:20:51.000-0800

**Comment by rbeckman-nextgen** [comment]
The only complaint I have is the error message received when attempting to use a resource that isn't configured for a particular connector. It wasn't immediately apparent to me that I forgot to configure this resource. I initially thought that my class wasn't exposing the method mentioned by the error or wasn't present in the jar file entirely.

Transformer error
ERROR MESSAGE: Error evaluating transformer
com.mirth.connect.server.MirthJavascriptTransformerException: 
CHANNEL:	Resources
CONNECTOR:	Destination 4
SCRIPT SOURCE:	TRANSFORMER
SOURCE CODE:	
25: function doScript() {
26: if (connectorMessage.getProcessedRawData() != null) { msg = new String(connectorMessage.getProcessedRawData()); } else { msg = new String(connectorMessage.getRawData()); } 
27: function doFilter() { phase[0] = 'filter'; return true; }function doTransform() { phase[0] = 'transformer'; logger = Packages.org.apache.log4j.Logger.getLogger(phase[0]);
28: 
29: 
30: logger.info("Draw " + com.custom.Draw.draw());
31: if ('xml' === typeof msg && msg.hasSimpleContent()) { msg = msg.toXMLString(); }if ('xml' === typeof tmp && tmp.hasSimpleContent()) { tmp = tmp.toXMLString(); }
32: }
33: if (doFilter() == true) { doTransform(); return true; } else { return false; } 
34: }
LINE NUMBER:	30
DETAILS:	TypeError: Cannot call property draw in object [JavaPackage com.custom.Draw]. It is not a function, it is "object".
	at 1cfbef84-3e12-438c-a303-0bd846d50f8f:30 (doTransform)
	at 1cfbef84-3e12-438c-a303-0bd846d50f8f:33 (doScript)
	at 1cfbef84-3e12-438c-a303-0bd846d50f8f:35
	at com.mirth.connect.server.transformers.JavaScriptFilterTransformer$FilterTransformerTask.call(JavaScriptFilterTransformer.java:153)
	at com.mirth.connect.server.transformers.JavaScriptFilterTransformer$FilterTransformerTask.call(JavaScriptFilterTransformer.java:1)
	at java.util.concurrent.FutureTask.run(FutureTask.java:262)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1145)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:615)
	at java.lang.Thread.run(Thread.java:744)


Imported Comment. Original Details:
Author: jaysenp
Created: 2015-01-20T12:25:30.000-0800

**Comment by rbeckman-nextgen** [comment]
Hmm, well that's the same error message you would get if you, say, misspelled a class name (like "Drwa"). At that point I think it's well within the realm of user-error.

Imported Comment. Original Details:
Author: narupley
Created: 2015-01-20T12:38:30.000-0800

**Comment by rbeckman-nextgen** [comment]
Closing since I failed to do so after completing testing.

Imported Comment. Original Details:
Author: jaysenp
Created: 2015-01-27T16:05:59.000-0800

**Comment by rbeckman-nextgen** [comment]
We'll have to be careful with this in the future because using custom classloaders can break singletons.
http://stackoverflow.com/questions/15322093/breaking-of-sinlgleton-by-two-different-class-loaders

I think in this case it should be fine since you're setting the parent classloader (not 100% sure on this though), but we should remember this if we decide to do more with classloaders in the future.

Imported Comment. Original Details:
Author: wayneh
Created: 2015-02-15T16:30:05.000-0800

**Comment by rbeckman-nextgen** [comment]
@Wayne Huang I ran into the singleton problem while using Oracle UCP Manager.I had to keep a reference to the singleton in a global map and reuse it.
http://docs.oracle.com/cd/E11882_01/java.112/e12265/manage.htm#BABGHEDJ

Also I am running into problems with classloading priority. Because you set parent classloader on each isolated classloader the priority will always be the parent classloader class and not the specific isolated and user specified classloader. Mirth parent classloader has a lot of classes alreadey loaded that user might want to use like different Oracle drivers or PDF Box version or Hapi libraries...
I think there should be a way of configuring classloading priorities like some application servers do. See jetty example here : http://www.eclipse.org/jetty/documentation/current/jetty-classloading.html 

Imported Comment. Original Details:
Author: hugosoares2
Created: 2015-11-09T18:08:56.000-0800
